openedx/edx-ora2
# feat: implement grading strategies for peer evaluations

**TL;DR -** This PR adds a new grading calculation strategy that can be configured in the assessment steps configurations for peer steps instead of the median calculation.

Currently, for steps that calculate a final score based on multiple scores (e.g. peer steps) do so using the median by default; the calculation is done [here](https://github.com/openedx/edx-ora2/blob/a2f652ac6aa1a0758bacc4832a1a52445229059e/openassessment/assessment/models/base.py#L491-L520). Using the median for this use case has its advantages. However, using it by default without reconfigurability options removes other use cases that can be done using the mean calculation for the final score instead. This also opens the door for other grading strategies mentioned in [this thread](https://openedx.slack.com/archives/C014VAUJUJ0/p1695241319300679).

That's why [this product proposal](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3891462181/Configurable+grading+strategy+for+peer+assessments) was created and later approved. 

**What changed?**

- The students' score is obtained before setting the score [here](https://github.com/openedx/edx-ora2/blob/a2f652ac6aa1a0758bacc4832a1a52445229059e/openassessment/workflow/models.py#L439-L444)
- Each step API has its own `get_score` function
- The peer step `get_score` gets the median score of all scores the student has at the moment of generation by using [get_assessments_median_scores](https://github.com/openedx/edx-ora2/blob/a2f652ac6aa1a0758bacc4832a1a52445229059e/openassessment/assessment/api/peer.py#L264-L267)
- The `get_assessments_median_scores` calls `get_median_scores_dict`, which calculates the final score for the student using the median.

So we can make the grading strategy configurable we:

Changed each reference to `get_assessments_median_scores` for a more generic `get_assessments_peer_scores_with_grading_strategy` which discriminates based on the workflow configuration which calculation to use. If the feature flag `ENABLE_ORA_PEER_CONFIGURABLE_GRADING` is off, then the default `get_assessments_median_scores` is used.

By default, the median calculations are used.

**Developer Checklist**


**Testing Instructions**

1. Clone this branch into your file system
2. Install edx-ora2 as an editable requirement (-e optional)
3. Now, add an ORA component to your course. 
4. Turn on the feature by adding: `FEATURES["ENABLE_ORA_PEER_CONFIGURABLE_GRADING"] = True`
5. Configure the steps as follows so these tests instructions make sense:
![image](https://github.com/openedx/edx-ora2/assets/64440265/1fee496b-95e8-45b5-879f-80e5d63d2c15)
6. Use just one rubric configured with these option points:
![image](https://github.com/eduNEXT/edx-ora2/assets/64440265/be9a23b7-4449-40bb-8009-64eb44a80657)
7. The median and the mean look alike when the dataset is not skewed. So, so we can check we're using the mean instead of the median, let's use the following scoring sequence:
Student 1 submits the ORA activity with a distinctive answer so we can distinguish it later
Student 2 scores Student 1 with 0
Student 3 scores Student 1 with 1
Student 4 scores Student 1 with 5
Now, the median is: 1 but the mean is: 2. So check that grading matches the mean value and the explanation says mean instead of median.
![image](https://github.com/eduNEXT/edx-ora2/assets/64440265/433444d9-134e-4311-b566-231a300d1872)

If you don't configure the mean grading then the default behavior is kept:
![image](https://github.com/eduNEXT/edx-ora2/assets/64440265/0f7a5285-cc7b-4c89-8afd-8df1c0f5eda2)


**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:
